src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py

```diff
@@ -99,6 +108,73 @@ def load_config(self):
         else:
             self.metadata_mgr = MetadataManager(self.fs, self.config_path, 0o640)
 
+    def _create(self, size, isolate_nspace, pool, mode, uid, gid, version, need_incarnation=False):
+        subvolume_type = SubvolumeBase.SUBVOLUME_TYPE_NORMAL
+        try:
+            initial_state = OpSm.get_init_state(subvolume_type)
+        except OpSmException as oe:
+            raise VolumeException(-errno.EINVAL, "subvolume creation failed: internal error")
+
+        # TODO: UUID reuse checks
+        incarnation = str(uuid.uuid4()).encode('utf-8')
+        subvol_path = os.path.join(self.base_path, incarnation)
+        try:
+            # create directory and set attributes
+            self.fs.mkdirs(subvol_path, mode)
+            self.set_attrs(subvol_path, size, isolate_nspace, pool, uid, gid)
+
+            # persist subvolume metadata
+            qpath = subvol_path.decode('utf-8')
+            # check if reincarnating a subvolume, to preserve older metadata
+            preserve = False
+            try:
+                self.metadata_mgr.refresh()
+                preserve = True
+            except MetadataMgrException as me:
+                if me.errno != -errno.ENOENT:
+                    raise
+                log.info("reincarnating subvolume with path: {0}".format(self.subvolname))
+
+            if not preserve:
+                if need_incarnation:
+                    self.init_config(version, subvolume_type, qpath, initial_state, incarnation)
+                else:
+                    self.init_config(version, subvolume_type, qpath, initial_state)
+            else:
+                self.metadata_mgr.update_global_section(MetadataManager.GLOBAL_META_KEY_TYPE, subvolume_type)
+                self.metadata_mgr.update_global_section(MetadataManager.GLOBAL_META_KEY_PATH, qpath)
+                self.metadata_mgr.update_global_section(MetadataManager.GLOBAL_META_KEY_STATE, initial_state)
+                self.metadata_mgr.update_global_section(MetadataManager.GLOBAL_META_KEY_INCARNATION, incarnation.decode('utf-8'))
+                self.metadata_mgr.flush(incarnation)
+        except (VolumeException, MetadataMgrException, cephfs.Error) as e:
+            try:
+                log.info("cleaning up subvolume with path: {0}".format(self.subvolname))
+                self.remove()
+            except VolumeException as ve:
+                log.info("failed to cleanup subvolume '{0}' ({1})".format(self.subvolname, ve))
+
+            if isinstance(e, MetadataMgrException):
+                log.error("metadata manager exception: {0}".format(e))
+                e = VolumeException(-errno.EINVAL, "exception in subvolume metadata")
+            elif isinstance(e, cephfs.Error):
+                e = VolumeException(-e.args[0], e.args[1])
+            raise e
+
+    def add_clone_source(self, volname, subvolume, snapname, flush=False):
```

Comment:
As most of the implementation is common across v1/v2 subvolume versions, the code would move to the subvolume base file, and where appropriate retain the base class method with the same name.